src/utils/fxa.ts

```diff
@@ -0,0 +1,351 @@
+  }
+}
+/* c8 ignore stop */
+
+// Not covered by tests; mostly side-effects. See test-coverage.md#mock-heavy
+/* c8 ignore start */
+async function deleteSubscription(bearerToken: string): Promise<boolean> {
+  try {
+    const subs = (await getSubscriptions(bearerToken)) ?? [];
+    let subscriptionId;
+    for (const sub of subs) {
+      if (
+        sub &&
+        sub.productId &&
+        sub.productId === AppConstants.PREMIUM_PRODUCT_ID
+      ) {
+        subscriptionId = sub.subscriptionId;
+      }
+    }
+    if (subscriptionId) {
+      const deleteUrl = `${AppConstants.OAUTH_ACCOUNT_URI}/oauth/subscriptions/active/${subscriptionId}`;
+      const response = await fetch(deleteUrl, {
+        method: "DELETE",
+        headers: {
+          Accept: "application/json",
+          Authorization: `Bearer ${bearerToken}`,
+        },
+      });
+      if (!response.ok) {
+        throw new Error(await response.text());
+      } else {
+        console.info(
+          `delete_fxa_subscription: success - ${JSON.stringify(await response.json())}`,
+        );
+      }
+    }
+    return true;
+  } catch (e) {
+    if (e instanceof Error) {
+      console.error("delete_fxa_subscription", { stack: e.stack });
+    }
+    return false;
+  }
+}
+/* c8 ignore stop */
+
+// Not covered by tests; mostly side-effects. See test-coverage.md#mock-heavy
+/* c8 ignore start */
+async function applyCoupon(
+  bearerToken: string,
+  couponCodeId: string,
+): Promise<void> {
+  try {
+    const subs = (await getSubscriptions(bearerToken)) ?? [];
+    let subscriptionId;
+    for (const sub of subs) {
+      if (
+        sub &&
+        sub.productId &&
+        sub.productId === AppConstants.PREMIUM_PRODUCT_ID
+      ) {
+        subscriptionId = sub.subscriptionId;
+      }
+    }
+    if (subscriptionId) {
+      const applyCouponUrl = `${AppConstants.OAUTH_ACCOUNT_URI}/oauth/subscriptions/coupon/apply`;
+      const response = await fetch(applyCouponUrl, {
+        method: "PUT",
+        headers: {
+          "Content-Type": "application/json",
+          Accept: "application/json",
+          Authorization: `Bearer ${bearerToken}`,
+        },
+        body: JSON.stringify({
+          promotionId: couponCodeId,
+          subscriptionId,
+        }),
+      });
+      if (!response.ok) {
+        const errMsg = await response.text();
+        console.error(`apply_coupon: failed - ${errMsg}`);
+        throw new Error(`apply_coupon: failed - ${errMsg}`);
+      } else {
+        console.info(
+          `apply_coupon: success - ${JSON.stringify(await response.json())}`,
+        );
+      }
+    }
+  } catch (e) {
+    if (e instanceof Error) {
+      console.error("apply_coupon", { stack: e.stack });
+    }
+    throw e;
+  }
+}
+/* c8 ignore stop */
+
+// TODO: Add unit test when changing this code:
+/* c8 ignore next 3 */
+function getSha1(email: crypto.BinaryLike) {
+  return crypto.createHash("sha1").update(email).digest("hex");
```

Comment:
I assume that this is part of the HIBP API contract and not something we can just change? (And not even sure we should, since IIRC it's not used for encryption.) Either way this isn't a regression, as this is just copying over code from `fxa.js`.

Reply:
Yeah, there's more documentation here: https://www.troyhunt.com/understanding-have-i-been-pwneds-use-of-sha-1-and-k-anonymity/